Mathlib/Algebra/Category/Ring/Under.lean

```diff
@@ -0,0 +1,159 @@
+    (fromAlgHom (g.comp f)) = fromAlgHom f ≫ fromAlgHom g :=
+  rfl
+
+lemma mkUnder_ext {A B : Type u} [CommRing A] [CommRing B] [Algebra R A] [Algebra R B]
+    {f g : mkUnder R A ⟶ mkUnder R B} (h : ∀ a : A, f.right a = g.right a) :
+    f = g := by
+  ext x
+  exact h x
+
+/-- Make an isomorphism in `Under R` from an algebra isomorphism. -/
+def fromAlgEquiv {A B : Type u} [CommRing A] [CommRing B] [Algebra R A] [Algebra R B]
+    (f : A ≃ₐ[R] B) :
+    mkUnder R A ≅ mkUnder R B where
+  hom := fromAlgHom f.toAlgHom
+  inv := fromAlgHom f.symm.toAlgHom
+  hom_inv_id := by
+    ext (a : (mkUnder R A).right)
+    simp
+  inv_hom_id := by
+    ext a
+    simp
+
+@[simp]
+lemma fromAlgEquiv_hom_right_apply {A B : Type u} [CommRing A] [CommRing B] [Algebra R A]
+    [Algebra R B] (f : A ≃ₐ[R] B) (a : A) :
+    (fromAlgEquiv f).hom.right a = f a := rfl
+
+@[simp]
+lemma fromAlgEquiv_inv_right_apply {A B : Type u} [CommRing A] [CommRing B] [Algebra R A]
+    [Algebra R B] (f : A ≃ₐ[R] B) (b : B) :
+    (fromAlgEquiv f).inv.right b = f.symm b := rfl
+
+@[simp]
+lemma fromAlgEquiv_trans {A B C : Type u} [CommRing A] [CommRing B] [CommRing C] [Algebra R A]
+    [Algebra R B] [Algebra R C] (f : A ≃ₐ[R] B) (g : B ≃ₐ[R] C) :
+    (fromAlgEquiv (f.trans g)) = fromAlgEquiv f ≪≫ fromAlgEquiv g :=
+  rfl
+
+variable [Algebra R S]
+
+variable (R S) in
+/-- The base change functor `A ↦ S ⊗[R] A`. -/
+@[simps! map_right]
+def tensorProd : Under R ⥤ Under S where
+  obj A := mkUnder S (S ⊗[R] A)
+  map f := fromAlgHom <| Algebra.TensorProduct.map (AlgHom.id S S) (toAlgHom f)
+  map_comp {X Y Z} f g := by simp [Algebra.TensorProduct.map_id_comp]
+
+variable (S) in
+/-- The natural isomorphism `S ⊗[R] A ≅ pushout A.hom (algebraMap R S)` in `Under S`. -/
+def tensorProdObjIsoPushoutObj (A : Under R) :
+    mkUnder S (S ⊗[R] A) ≅ (Under.pushout (algebraMap R S)).obj A :=
+  Under.isoMk (CommRingCat.isPushout_tensorProduct R S A).flip.isoPushout <| by
+    simp only [Functor.const_obj_obj, Under.pushout_obj, Functor.id_obj, Under.mk_right,
+      mkUnder_hom, AlgHom.toRingHom_eq_coe, IsPushout.inr_isoPushout_hom, Under.mk_hom]
+    rfl
+
+@[reassoc (attr := simp)]
+lemma pushout_inl_tensorProdObjIsoPushoutObj_inv_right (A : Under R) :
+    pushout.inl A.hom (algebraMap R S) ≫ (tensorProdObjIsoPushoutObj S A).inv.right =
+      (CommRingCat.ofHom <| Algebra.TensorProduct.includeRight.toRingHom) := by
+  simp [tensorProdObjIsoPushoutObj]
+
+@[reassoc (attr := simp)]
+lemma pushout_inr_tensorProdObjIsoPushoutObj_inv_right (A : Under R) :
+    pushout.inr A.hom (algebraMap R S) ≫
+      (tensorProdObjIsoPushoutObj S A).inv.right =
+      (CommRingCat.ofHom <| Algebra.TensorProduct.includeLeftRingHom) := by
+  simp [tensorProdObjIsoPushoutObj]
+
+/-- `A ↦ S ⊗[R] A` is naturally isomorphic to `A ↦ pushout A.hom (algebraMap R S)`. -/
+def tensorProdIsoPushout : tensorProd R S ≅ Under.pushout (algebraMap R S) :=
```

Comment:
```suggestion
@[simps!]
def tensorProdIsoPushout : tensorProd R S ≅ Under.pushout (algebraMap R S) :=
```
Maybe?

Reply:
This produces horrible lemmas, I manually added an `_app` lemma.